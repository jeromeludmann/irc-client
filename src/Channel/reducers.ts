import { combineReducers } from "redux";
import {
  CommandInputState,
  commandInputReducer,
} from "@app/CommandInput/reducers";
import { ChannelScope } from "@app/types";

// Name

type NameState = string;
const nameInitialState: NameState = "";
const nameReducer = (state = nameInitialState): NameState => state;

// Messages

type MessagesState = string[];

type MessagesPayload = ChannelScope & { value: string };

interface MessagesAction {
  type: "COMMAND_SENT";
  payload: MessagesPayload;
}

const messagesInitialState: MessagesState = [];

const messagesReducer = (
  state = messagesInitialState,
  { type, payload }: MessagesAction,
): MessagesState => {
  switch (type) {
    case "COMMAND_SENT":
      return [...state, payload.value];
    default:
      return state;
  }
};

export interface ChannelState {
  name: NameState;
  messages: MessagesState;
  commandInput: CommandInputState;
}

export const channelReducer = combineReducers<ChannelState>({
  name: nameReducer,
  messages: messagesReducer,
  commandInput: commandInputReducer,
});
