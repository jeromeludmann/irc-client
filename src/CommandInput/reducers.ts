import { combineReducers } from "redux";

// Value

type ValueState = string;

type ValueAction =
  | {
      type: "COMMAND_CHANGED";
      payload: { value: string };
    }
  | {
      type: "COMMAND_SENT";
      payload: {};
    };

const valueInitialState: ValueState = "";

const value = (state = valueInitialState, action: ValueAction): ValueState => {
  switch (action.type) {
    case "COMMAND_CHANGED":
      return action.payload.value;
    case "COMMAND_SENT":
      return "";
    default:
      return state;
  }
};

// History

type HistoryState = string[];
const historyInitialState: HistoryState = [];
const history = (state = historyInitialState): HistoryState => state;

// CommandInput

export type CommandInputState = {
  value: ValueState;
  history: HistoryState;
};

export const commandInputReducer = combineReducers<CommandInputState>({
  value,
  history,
});
