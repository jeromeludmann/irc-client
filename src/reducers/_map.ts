import { Action } from "redux";

export const mapReducer = <S, E = void>(map: {
  [action: string]: (state: S, action: Action, extraStates?: E) => S;
}) => (state: S, action: Action, extraStates?: E): S =>
  map.hasOwnProperty(action.type)
    ? map[action.type](state, action, extraStates)
    : state;
