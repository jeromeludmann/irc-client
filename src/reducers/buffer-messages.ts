import {
  CONNECTION_FAILED,
  ConnectionFailedAction,
  RAW_MESSAGES_RECEIVED,
  RawMessagesReceivedAction,
} from "@app/actions/network";
import {
  JoinAction,
  ServerNoticeAction,
  ServerPingAction,
  ChannelPrivmsgAction,
  UserPrivmsgAction,
  JOIN,
  SERVER_PING,
  CHANNEL_PRIVMSG,
  USER_PRIVMSG,
  SERVER_NOTICE,
} from "@app/actions/message-in";

export type MessageListState = string[];

export type MessageListAction =
  | JoinAction
  | ServerNoticeAction
  | ServerPingAction
  | ChannelPrivmsgAction
  | UserPrivmsgAction
  | ConnectionFailedAction
  | RawMessagesReceivedAction;

export const messagesInitialState: MessageListState = [];

export default (
  messages = messagesInitialState,
  action: MessageListAction,
): MessageListState => {
  switch (action.type) {
    case JOIN: {
      const { user, channel } = action.payload;
      const msg = `${user.nick} has joined ${channel}`;
      return [...messages, msg];
    }
    case SERVER_PING: {
      return [...messages, "Ping? (Pong!)"];
    }
    case CHANNEL_PRIVMSG:
    case USER_PRIVMSG: {
      const { user, text } = action.payload;
      const msg = `${user.nick}: ${text}`;
      return [...messages, msg];
    }
    case SERVER_NOTICE: {
      return [...messages, action.payload.text];
    }
    case CONNECTION_FAILED: {
      return [...messages, action.payload.message];
    }
    case RAW_MESSAGES_RECEIVED: {
      return [...messages, ...action.payload.messages];
    }
    default:
      return messages;
  }
};
