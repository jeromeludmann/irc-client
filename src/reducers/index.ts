import {
  RouteState,
  routeInitialState,
  reduceRoute,
} from "@app/reducers/route";
import {
  ServersState,
  serversInitialState,
  reduceServers,
} from "@app/reducers/servers";
import { RoutedAction } from "@app/Route";
import { Reducer } from "redux";

export interface AppState {
  readonly servers: ServersState;
  readonly route: RouteState;
}

export const rootInitialState = {
  servers: serversInitialState,
  route: routeInitialState,
};

export const reduce: Reducer<AppState, RoutedAction> = (
  state = rootInitialState,
  action,
) => ({
  servers: reduceServers(state.servers, action, { route: state.route }),
  route: reduceRoute(state.route, action, { servers: state.servers }),
});
