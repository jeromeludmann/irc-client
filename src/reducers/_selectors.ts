import { createSelector } from "reselect";
import { RootState } from "@app/reducers";
import { ServersState } from "@app/reducers/servers";
import { RouteState } from "@app/reducers/route";

export const selectServers = ({ servers }: RootState): ServersState => servers;

export const selectRoute = ({ route }: RootState): RouteState => route;

export const selectServer = createSelector(
  selectServers,
  selectRoute,
  (servers, { serverKey }) => servers[serverKey],
);
