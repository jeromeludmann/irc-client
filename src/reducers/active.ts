import { Route, STATUS_WINDOW } from "@app/Route";
import { SwitchWindowAction, SWITCH_WINDOW } from "@app/actions/ui-window";
import { JoinAction, JOIN } from "@app/actions/message-in";
import { UserState } from "@app/reducers/user";

export type ActiveWindowState = Route;

export type ActiveWindowAction = JoinAction | SwitchWindowAction;

export const activeInitialState: ActiveWindowState = {
  server: "serverKey", // TODO multi servers
  window: STATUS_WINDOW,
};

type ExtraParams = {
  user: UserState;
};

export default (
  active = activeInitialState,
  action: ActiveWindowAction,
  { user }: ExtraParams,
): ActiveWindowState => {
  switch (action.type) {
    case JOIN:
      return action.payload.user.nick === user.nick ? action.route : active;

    case SWITCH_WINDOW:
      return action.route;

    default:
      return active;
  }
};
