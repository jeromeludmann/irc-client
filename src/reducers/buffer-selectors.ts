import { createSelector } from "reselect";
import { selectWindow } from "@app/reducers/server-selectors";

export const selectMessages = createSelector(
  selectWindow,
  window => window.messages,
);

export const selectInput = createSelector(selectWindow, window => window.input);
