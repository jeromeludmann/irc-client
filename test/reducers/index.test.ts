import { reduce, rootInitialState } from "@app/reducers";

describe("root reducer", () => {
  test("initializing without state", () => {
    expect(
      reduce(undefined, {
        type: "",
        route: { serverKey: "", channelKey: "" },
      }),
    ).toMatchSnapshot();
  });
  test("initializing with state", () => {
    expect(
      reduce(rootInitialState, {
        type: "",
        route: { serverKey: "", channelKey: "" },
      }),
    ).toMatchSnapshot();
  });
});
