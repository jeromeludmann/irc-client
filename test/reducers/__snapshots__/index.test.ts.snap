// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`root reducer initializing with state 1`] = `
Object {
  "route": Object {
    "channelKey": "@status",
    "serverKey": "",
  },
  "servers": Object {},
}
`;

exports[`root reducer initializing without state 1`] = `
Object {
  "route": Object {
    "channelKey": "@status",
    "serverKey": "",
  },
  "servers": Object {},
}
`;
